h2o-genmodel/src/main/java/hex/genmodel/algos/ensemble/StackedEnsembleMojoModel.java

```diff
@@ -2,43 +2,90 @@
         double[] basePredsRow = new double[preds.length];
         if(_nclasses > 2) { //Multinomial
             int k = 0;
             basePreds = new double[_baseModelNum * _nclasses]; //Proper allocation for multinomial ensemble (class probabilities per base model)
             for(int i = 0; i < _baseModelNum; ++i){
                 for(int j = 0; j < _nclasses; ++j){
-                    basePreds[k] = _baseModels[i].score0(row, basePredsRow)[j+1];
+                    basePreds[k] = _baseModels[i]._mojoModel.score0(_baseModels[i].remapRow(row, _precalculatedMappings),
+                            basePredsRow)[j + 1];
                     k++;
                 }
             }
         }else if(_nclasses == 2){ //Binomial
             for(int i = 0; i < _baseModelNum; ++i) {
-                _baseModels[i].score0(row, basePredsRow);
+                _baseModels[i]._mojoModel.score0(_baseModels[i].remapRow(row, _precalculatedMappings), basePredsRow);
                 basePreds[i] = basePredsRow[2];
             }
         }else{ //Regression
             for(int i = 0; i < _baseModelNum; ++i) { //Regression
-                _baseModels[i].score0(row, basePredsRow);
+                _baseModels[i]._mojoModel.score0(_baseModels[i].remapRow(row, _precalculatedMappings), basePredsRow);
                 basePreds[i] = basePredsRow[0];
             }
         }
         _metaLearner.score0(basePreds, preds);
         return preds;
     }
 
+    /**
+     * In stacked ensembles, multiple models may appear. Problem with multiple models present is possibly different
+     * internal order of features. Therefore, the scored row's values must be re-mapped to the internal order of each
+     * model.
+     */
+    static class StackedEnsembleMojoSubModel {
+
+        final MojoModel _mojoModel;
+        final int[] _mapping; // Mapping. If null, no mapping is required.
+        final int _mappingHashIndex;
+
+        public StackedEnsembleMojoSubModel(MojoModel mojoModel, int[] mapping, int mappingHash) {
+            _mojoModel = mojoModel;
+            _mapping = mapping;
+            _mappingHashIndex = mappingHash;
+        }
+
+        /**
+         * Returns a new array represeting row values re-mapped to order given by the underlying submodel.
+         * Order of columns in the row may remain the same, yet a new instance of double[] is returned all the time.
+         *
+         * @param row Row to re-map
+         * @return A new instance of double[] with values re-mapped to order given by the underlying submodel.
+         */
+        public double[] remapRow(final double[] row, final double[][] _precalculatedMappings) {
+            if (_precalculatedMappings[_mappingHashIndex] != null) {
+                return Arrays.copyOf(_precalculatedMappings[_mappingHashIndex], _precalculatedMappings[_mappingHashIndex].length);
```

Comment:
you are still making a copy of the input array and touching every single element, so this function has a linear complexity no matter if you hit the cache or not